src/RpcServer/RpcServer.Blockchain.cs

```diff
@@ -105,20 +105,16 @@ protected virtual JObject GetBlockHeader(JArray _params)
         protected virtual JObject GetContractState(JArray _params)
         {
             using SnapshotView snapshot = Blockchain.Singleton.GetSnapshot();
-            UInt160 script_hash = ToScriptHash(_params[0].AsString());
+            UInt160 script_hash = ToScriptHash(snapshot, _params[0].AsString());
             ContractState contract = NativeContract.ContractManagement.GetContract(snapshot, script_hash);
             return contract?.ToJson() ?? throw new RpcException(-100, "Unknown contract");
         }
 
-        private static UInt160 ToScriptHash(string keyword)
+        private static UInt160 ToScriptHash(SnapshotView snapshot, string keyword)
         {
-            foreach (var native in NativeContract.Contracts)
-            {
-                if (keyword.Equals(native.Name, StringComparison.InvariantCultureIgnoreCase) || keyword == native.Id.ToString())
-                    return native.Hash;
-            }
-
-            return UInt160.Parse(keyword);
+            return NativeContract.Contracts.FirstOrDefault(p => keyword.Equals(p.Name, StringComparison.InvariantCultureIgnoreCase) || keyword == p.Id.ToString())?.Hash ??
+                   NativeContract.ContractManagement.ListContracts(snapshot).FirstOrDefault(p => keyword.Equals(p.Manifest.Name, StringComparison.InvariantCultureIgnoreCase) || keyword == p.Id.ToString())?.Hash ??
```

Comment:
It will fetch all the smart contracts, we should test it with at least 100k smart contracts. In order to avoid a possible Denial of Service

Reply:
> It will fetch all the smart contracts, we should test it with at least 100k smart contracts. In order to avoid a possible Denial of Service

Maybe we could move `getcontractstate` to somewhere like ApplicationLogs, then it could get contract state indenpently from the beginning just with a little adjustment on that plugin?

Reply:
@devhawk  what do you think? if it is possible to cause Denial of Service, we need to reconsider what the API benefits. Otherwise there is no conclusion on this.

Reply:
I understand the need for a stable contract identifier, but the choice of generating it from a combination of the original deployed contract hash + the original contract deployer address is not very developer friendly. I think there needs to be a mechanism for developers to be able to retrieve information about their deployed contracts w/o having to understand the algorithm that generates the identification hash. 

If we don't make this change to getContractState, another option could be to introduce a new contract search RPC method that returns an array of all contacts identification hashes (along with associated metadata) that matches a given name.